cdc/sink/dmlsink/mq/dmlproducer/pulsar_dml_producer.go

```diff
@@ -0,0 +1,312 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dmlproducer
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+	"time"
+
+	"github.com/apache/pulsar-client-go/pulsar"
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/cdc/sink/metrics/mq"
+	"github.com/pingcap/tiflow/pkg/config"
+	cerror "github.com/pingcap/tiflow/pkg/errors"
+	"github.com/pingcap/tiflow/pkg/sink/codec/common"
+	"go.uber.org/zap"
+)
+
+var _ DMLProducer = (*pulsarDMLProducer)(nil)
+
+// pulsarDMLProducer is used to send messages to kafka.
+type pulsarDMLProducer struct {
+	// id indicates which processor (changefeed) this sink belongs to.
+	id model.ChangeFeedID
+	// We hold the client to make close operation faster.
+	// Please see the comment of Close().
+	client pulsar.Client
+	// producers is used to send messages to kafka asynchronously.
+	// support multiple topics
+	producers *lru.Cache
+
+	// closedMu is used to protect `closed`.
+	// We need to ensure that closed producers are never written to.
+	closedMu sync.RWMutex
+	// closed is used to indicate whether the producer is closed.
+	// We also use it to guard against double closes.
+	closed bool
+	// closedChan is used to notify the run loop to exit.
+	closedChan chan struct{}
```

Comment:
there is no need to maintain a `closedChan` to exit the `loop`, I think this can be solved by using `ctx.Cancel()`.